pkgs/dart_model/lib/src/json_buffer/typed_map.dart

```diff
@@ -0,0 +1,353 @@
+        : (v0 == null ? 0 : schema.typeAtIndex(0).sizeInBytes) +
+            (v1 == null ? 0 : schema.typeAtIndex(1).sizeInBytes) +
+            (v2 == null ? 0 : schema.typeAtIndex(2).sizeInBytes) +
+            (v3 == null ? 0 : schema.typeAtIndex(3).sizeInBytes) +
+            (v4 == null ? 0 : schema.typeAtIndex(4).sizeInBytes) +
+            (v5 == null ? 0 : schema.typeAtIndex(5).sizeInBytes) +
+            (v6 == null ? 0 : schema.typeAtIndex(6).sizeInBytes) +
+            (v7 == null ? 0 : schema.typeAtIndex(7).sizeInBytes);
+
+    _reserve(_pointerSize + (filled ? 0 : schema.fieldSetSize) + size);
+
+    if (!filled) {
+      _set(
+          pointer + _pointerSize,
+          (v0 == null ? 0 : 1) +
+              (v1 == null ? 0 : 2) +
+              (v2 == null ? 0 : 4) +
+              (v3 == null ? 0 : 8) +
+              (v4 == null ? 0 : 16) +
+              (v5 == null ? 0 : 32) +
+              (v6 == null ? 0 : 64) +
+              (v7 == null ? 0 : 128));
+    }
+
+    _writePointer(pointer, schemaPointer);
+
+    var offset = 0;
+
+    void addValue(int index, Object? value) {
+      explanations?.push('addValue $index $value');
+
+      if (schema.isAllBooleans) {
+        final valuePointer = pointer +
+            _pointerSize +
+            (filled ? 0 : schema.fieldSetSize) +
+            (offset ~/ 8);
+        final bitIndex = offset % 8;
+        _setBit(valuePointer, bitIndex, value as bool);
+        offset++;
+      } else {
+        final valuePointer = pointer +
+            _pointerSize +
+            (filled ? 0 : schema.fieldSetSize) +
+            offset;
+        final valueType = schema.typeAtIndex(index);
+        _writeAnyOfType(valueType, valuePointer, value);
+        offset += valueType.sizeInBytes;
+      }
+      explanations?.pop();
+    }
+
+    if (v0 != null) addValue(0, v0);
+    if (v1 != null) addValue(1, v1);
+    if (v2 != null) addValue(2, v2);
+    if (v3 != null) addValue(3, v3);
+    if (v4 != null) addValue(4, v4);
+    if (v5 != null) addValue(5, v5);
+    if (v6 != null) addValue(6, v6);
+    if (v7 != null) addValue(7, v7);
+
+    explanations?.pop();
+    return _TypedMap(this, pointer);
+  }
+
+  Pointer _pointerToTypedMap(_TypedMap map) {
+    _checkTypedMapOwnership(map);
+    return map._pointer;
+  }
+
+  /// Throws if [map is backed by a different buffer to `this`.
+  void _checkTypedMapOwnership(_TypedMap map) {
+    if (map._buffer != this) {
+      throw UnsupportedError('Maps created with `addGrowableMap` can only '
+          'be added to the JsonBufferBuilder instance that created them.');
+    }
+  }
+
+  Map<String, Object?> readTypedMap(Pointer pointer) {
+    return _TypedMap(this, pointer);
+  }
+}
+
+class _TypedMap
+    with MapMixin<String, Object?>, _EntryMapMixin
+    implements Map<String, Object?> {
+  final JsonBufferBuilder _buffer;
+  final Pointer _pointer;
+  final Pointer _schemaPointer;
+  late final TypedMapSchema _schema =
```

Comment:
nit: make these non-late and add a factory constructor

Reply:
These are intentionally late:

In analyzer+CFE where we are creating extension types "write only", i.e. immediately adding to some other map without keeping a reference, these reads from the buffer are not needed. The type is just serving to pass around the `_pointer` for use as a value by someone else.

Added a comment :)